Assets/HoloToolkit/Utilities/Scripts/Singleton.cs

```diff
@@ -60,45 +34,46 @@ public static bool IsInitialized
         }
 
         /// <summary>
-        /// Base Awake method that sets the Singleton's unique instance.
-        /// Called by Unity when initializing a MonoBehaviour.
-        /// Scripts that extend Singleton should be sure to call base.Awake() to ensure the
-        /// static Instance reference is properly created.
+        /// Base awake method that sets the singleton's unique instance.
         /// </summary>
         protected virtual void Awake()
         {
-            if (IsInitialized && instance != this)
+            if (instance != null)
             {
-                if (Application.isEditor)
-                {
-                    DestroyImmediate(this);
-                }
-                else
-                {
-                    Destroy(this);
-                }
-
-                Debug.LogErrorFormat("Trying to instantiate a second instance of singleton class {0}. Additional Instance was destroyed", GetType().Name);
+                Debug.LogErrorFormat("Trying to instantiate a second instance of singleton class {0}", GetType().Name);
             }
-            else if (!IsInitialized)
+            else
             {
                 instance = (T)this;
+                actionsWaitingForInitialization.ForEach(action => action());
+                actionsWaitingForInitialization = null;
             }
         }
 
-        /// <summary>
-        /// Base OnDestroy method that destroys the Singleton's unique instance.
-        /// Called by Unity when destroying a MonoBehaviour. Scripts that extend
-        /// Singleton should be sure to call base.OnDestroy() to ensure the
-        /// underlying static Instance reference is properly cleaned up.
-        /// </summary>
         protected virtual void OnDestroy()
         {
             if (instance == this)
             {
                 instance = null;
-                searchForInstance = true;
+            }
+        }
+
+        /// <summary>
+        /// Performs an action when the singleton is initialized.
+        /// If the singleton is already initialized, the action will be called instantanly.
+        /// Actions can be registered multiple times. In this case, the actions will also be called multiple times.
+        /// </summary>
+        /// <param name="action">Action to perform when the singleton is initialized.</param>
+        public static void WhenInitialized(Action action)
+        {
+            if (IsInitialized)
+            {
+                action();
+            }
+            else
+            {
+                actionsWaitingForInitialization.Add(action);
```

Comment:
How is this used in the rest of the Singleton instances.  Could you explain the _flow_ of events?